src/incoming_stream.rs

```diff
@@ -0,0 +1,306 @@
+use std::fmt;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use asynchronous_codec::FramedRead;
+use cid::CidGeneric;
+use fnv::FnvHashMap;
+use futures::future::{BoxFuture, Fuse, FusedFuture};
+use futures::{FutureExt, StreamExt};
+use tracing::error;
+
+use crate::cid_prefix::CidPrefix;
+use crate::message::Codec;
+use crate::multihasher::{MultihasherError, MultihasherTable};
+use crate::proto::message::mod_Message::BlockPresenceType;
+use crate::proto::message::mod_Message::Wantlist;
+use crate::proto::message::Message;
+
+/// Stream that reads `Message` and converts it to `IncomingMessage`.
+///
+/// On fatal errors `None` is returned which instruct `SelectAll` to drop the stream.
+pub(crate) struct IncomingStream<const S: usize> {
+    multihasher: Arc<MultihasherTable<S>>,
+    stream: FramedRead<libp2p_swarm::Stream, Codec>,
+    processing: Fuse<BoxFuture<'static, Option<IncomingMessage<S>>>>,
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct ClientMessage<const S: usize> {
+    pub(crate) block_presences: FnvHashMap<CidGeneric<S>, BlockPresenceType>,
+    pub(crate) blocks: FnvHashMap<CidGeneric<S>, Vec<u8>>,
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct ServerMessage {
+    pub(crate) wantlist: Wantlist,
+}
+
+#[derive(Debug, Default)]
+#[doc(hidden)]
+pub struct IncomingMessage<const S: usize> {
+    pub(crate) client: Option<ClientMessage<S>>,
+    pub(crate) server: Option<ServerMessage>,
+}
+
+impl<const S: usize> IncomingStream<S> {
+    pub(crate) fn new(stream: libp2p_swarm::Stream, multihasher: Arc<MultihasherTable<S>>) -> Self {
+        IncomingStream {
+            multihasher,
+            stream: FramedRead::new(stream, Codec),
+            processing: Fuse::terminated(),
+        }
+    }
+}
+
+impl<const S: usize> fmt::Debug for IncomingStream<S> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("IncomingStream { .. }")
+    }
+}
+
+impl<const S: usize> futures::Stream for IncomingStream<S> {
+    type Item = IncomingMessage<S>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<IncomingMessage<S>>> {
+        loop {
+            // If processing future is activated then poll it and return
+            // its result.
+            if !self.processing.is_terminated() {
+                match self.processing.poll_unpin(cx) {
+                    Poll::Ready(Some(msg)) => {
+                        // There is no need to forward an empty message
+                        if msg.client.is_some() || msg.server.is_some() {
```

Comment:
I'm not sure about this, because it is not forward compatible. For example if a newer bitswap version adds a new field which we do not decode, it would slash the stream instead of ignoring that message.